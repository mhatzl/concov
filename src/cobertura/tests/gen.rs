#[test]
fn gen_v4_read() {
    let content = include_str!("gen.xml");

    let cobertura = quick_xml::de::from_str::<crate::cobertura::schema::Coverage>(content).unwrap();

    insta::assert_ron_snapshot!(cobertura);
}

#[test]
fn gen_loose_read() {
    let content = include_str!("gen.xml");

    let cobertura =
        quick_xml::de::from_str::<crate::cobertura::schema_loose::Coverage>(content).unwrap();

    insta::assert_ron_snapshot!(cobertura);
}

#[test]
fn roundtrip() {
    let content = include_str!("gen.xml");
    let cobertura = quick_xml::de::from_str::<crate::cobertura::schema::Coverage>(content).unwrap();

    let ser_cobertura = cobertura.to_string();

    // removes whitespace formatting and possible x.0 for decimal representation
    let stripped_content = content
        .replace([' ', '\r', '\n', '\t'], "")
        .replace(".0\"", "\"");
    let stripped_cobertura = ser_cobertura
        .replace([' ', '\r', '\n', '\t'], "")
        .replace(".0\"", "\"");

    assert_eq!(
        stripped_cobertura, stripped_content,
        "Roundtrip cobertura doesn't macht original."
    );
}
