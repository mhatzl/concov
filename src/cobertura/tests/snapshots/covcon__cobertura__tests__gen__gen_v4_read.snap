---
source: src/cobertura/tests/gen.rs
expression: cobertura
---
coverage(
  r#@line-rate: 0.85,
  r#@branch-rate: 0.75,
  r#@lines-covered: 85,
  r#@lines-valid: 100,
  r#@branches-covered: 15,
  r#@branches-valid: 20,
  r#@complexity: 1.5,
  r#@version: "2.1.1",
  r#@timestamp: "1702139945",
  sources: Some(Sources(
    source: [
      "/path/to/source1",
      "/path/to/source2",
    ],
  )),
  packages: Packages(
    package: [
      Package(
        r#@name: "com.example.mypackage1",
        r#@line-rate: 0.9,
        r#@branch-rate: 0.8,
        r#@complexity: 1.2,
        classes: Classes(
          class: [
            Class(
              r#@name: "MyClass1",
              r#@filename: "com/example/mypackage1/MyClass1.java",
              r#@line-rate: 0.9,
              r#@branch-rate: 0.8,
              r#@complexity: 1.1,
              methods: Methods(
                method: [
                  Method(
                    r#@name: "method1",
                    r#@signature: "(I)V",
                    r#@line-rate: 1.0,
                    r#@branch-rate: 1.0,
                    r#@complexity: 1.0,
                    lines: Lines(
                      line: [
                        Line(
                          r#@number: 10,
                          r#@hits: 10,
                          r#@branch: false,
                        ),
                        Line(
                          r#@number: 15,
                          r#@hits: 5,
                          r#@branch: true,
                          r#@condition-coverage: Some("50%"),
                          conditions: Some(Conditions(
                            condition: [
                              Condition(
                                r#@number: 0,
                                r#@type: "jump",
                                r#@coverage: "50%",
                              ),
                              Condition(
                                r#@number: 1,
                                r#@type: "switch",
                                r#@coverage: "100%",
                              ),
                            ],
                          )),
                        ),
                      ],
                    ),
                  ),
                  Method(
                    r#@name: "method2",
                    r#@signature: "(Ljava/lang/String;)V",
                    r#@line-rate: 0.8,
                    r#@branch-rate: 0.6,
                    r#@complexity: 1.2,
                    lines: Lines(
                      line: [
                        Line(
                          r#@number: 20,
                          r#@hits: 3,
                          r#@branch: false,
                        ),
                        Line(
                          r#@number: 25,
                          r#@hits: 0,
                          r#@branch: true,
                          r#@condition-coverage: Some("0%"),
                          conditions: Some(Conditions(
                            condition: [
                              Condition(
                                r#@number: 0,
                                r#@type: "jump",
                                r#@coverage: "0%",
                              ),
                            ],
                          )),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
              lines: Lines(
                line: [
                  Line(
                    r#@number: 10,
                    r#@hits: 10,
                    r#@branch: false,
                  ),
                  Line(
                    r#@number: 15,
                    r#@hits: 5,
                    r#@branch: true,
                    r#@condition-coverage: Some("50%"),
                    conditions: Some(Conditions(
                      condition: [
                        Condition(
                          r#@number: 0,
                          r#@type: "jump",
                          r#@coverage: "50%",
                        ),
                        Condition(
                          r#@number: 1,
                          r#@type: "switch",
                          r#@coverage: "100%",
                        ),
                      ],
                    )),
                  ),
                ],
              ),
            ),
            Class(
              r#@name: "MyClass2",
              r#@filename: "com/example/mypackage1/MyClass2.java",
              r#@line-rate: 0.8,
              r#@branch-rate: 0.7,
              r#@complexity: 1.3,
              methods: Methods(
                method: [
                  Method(
                    r#@name: "method1",
                    r#@signature: "()V",
                    r#@line-rate: 1.0,
                    r#@branch-rate: 0.8,
                    r#@complexity: 1.1,
                    lines: Lines(
                      line: [
                        Line(
                          r#@number: 30,
                          r#@hits: 8,
                          r#@branch: false,
                        ),
                        Line(
                          r#@number: 35,
                          r#@hits: 2,
                          r#@branch: true,
                          r#@condition-coverage: Some("66%"),
                          conditions: Some(Conditions(
                            condition: [
                              Condition(
                                r#@number: 0,
                                r#@type: "jump",
                                r#@coverage: "66%",
                              ),
                            ],
                          )),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
              lines: Lines(
                line: [
                  Line(
                    r#@number: 30,
                    r#@hits: 8,
                    r#@branch: false,
                  ),
                  Line(
                    r#@number: 35,
                    r#@hits: 2,
                    r#@branch: true,
                    r#@condition-coverage: Some("66%"),
                    conditions: Some(Conditions(
                      condition: [
                        Condition(
                          r#@number: 0,
                          r#@type: "jump",
                          r#@coverage: "66%",
                        ),
                      ],
                    )),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
      Package(
        r#@name: "com.example.mypackage2",
        r#@line-rate: 0.85,
        r#@branch-rate: 0.75,
        r#@complexity: 1.4,
        classes: Classes(
          class: [
            Class(
              r#@name: "AnotherClass",
              r#@filename: "com/example/mypackage2/AnotherClass.java",
              r#@line-rate: 0.85,
              r#@branch-rate: 0.75,
              r#@complexity: 1.4,
              methods: Methods(
                method: [
                  Method(
                    r#@name: "anotherMethod",
                    r#@signature: "()I",
                    r#@line-rate: 1.0,
                    r#@branch-rate: 1.0,
                    r#@complexity: 1.0,
                    lines: Lines(
                      line: [
                        Line(
                          r#@number: 40,
                          r#@hits: 10,
                          r#@branch: false,
                        ),
                        Line(
                          r#@number: 45,
                          r#@hits: 7,
                          r#@branch: true,
                          r#@condition-coverage: Some("75%"),
                          conditions: Some(Conditions(
                            condition: [
                              Condition(
                                r#@number: 0,
                                r#@type: "jump",
                                r#@coverage: "75%",
                              ),
                            ],
                          )),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
              lines: Lines(
                line: [
                  Line(
                    r#@number: 40,
                    r#@hits: 10,
                    r#@branch: false,
                  ),
                  Line(
                    r#@number: 45,
                    r#@hits: 7,
                    r#@branch: true,
                    r#@condition-coverage: Some("75%"),
                    conditions: Some(Conditions(
                      condition: [
                        Condition(
                          r#@number: 0,
                          r#@type: "jump",
                          r#@coverage: "75%",
                        ),
                      ],
                    )),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    ],
  ),
)
